[tensorflow] Handle in-place array element initialization when decoding _allocateUninitialized.

dingobye opened the pull request:
When it comes to literal arrays that have some element(s) constructed in place rather than copied from a top level value, `ConstExprEvaluator` fails to decode such arrays, because the decoder requires every element to correspond to a top level value.

~~This patches tries to add stack variables to temporarily hold those element values and then store into the array. The code after this transformation can thus be properly handled by the decoder.~~

This updated patches adds supports to handle in-place element initilization. This is better compared to the previous solution which may change IR when decoding arrays.

Resolves [SR-8399](https://bugs.swift.org/browse/SR-8399).

dingobye committed:
[tensorflow] Generate top level stack variables to hold
array elements when decoding _allocateUninitialized.

When it comes to literal arrays that have some element(s)
constructed in place rather than copied from a top level
value, it fails to decode such arrays. This patches tries
to add stack variables to temporarily hold those element
values and then store into the array. The code after this
transformation can thus be properly handled by the decoder.

dingobye commented:
@dan-zheng Hi Dan, would you help review this, please?

dan-zheng reviewed (approved):
Great work! 😄

dan-zheng commented:
@swift-ci Please test tensorflow

dingobye commented:
Thank you for the review, Dan!

swift-ci [failed in a SourceKit test](https://ci-external.swift.org/job/swift-PR-TensorFlow-Linux/534/), which seems to be irrelevant to this PR. I cannot reproduce this failure locally. Should we wait for a while and fire a test again?

dan-zheng commented:
@swift-ci Please test tensorflow

mhong commented:
@dingobye, thank you for attempting the fix! I wonder if we could instead enhance ConstExprEvaluator to const-eval SIL code below (in particular the `SignedInteger<>.init<A>(_:)` call), so that an array like `[4, Int32(1)]` gets "canonicalized" into `[4, 1]`.

This will also help with other use cases of const-eval, such as `#assert`. (https://github.com/apple/swift/blob/tensorflow/test/SILOptimizer/pound_assert.swift). 

+@lattner for Chris' input.

```
  // function_ref _allocateUninitializedArray<A>(_:)
  %25 = function_ref @$Ss27_allocateUninitializedArrayySayxG_BptBwlF : $@convention(thin) <τ_0_0> (Builtin.Word) -> (@owned Array<τ_0_0>, Builtin.RawPointer) // user: %26
  %26 = apply %25<Int32>(%24) : $@convention(thin) <τ_0_0> (Builtin.Word) -> (@owned Array<τ_0_0>, Builtin.RawPointer) // users: %30, %27, %29
  %27 = tuple_extract %26 : $(Array<Int32>, Builtin.RawPointer), 0 // users: %47, %28
  retain_value %27 : $Array<Int32>                // id: %28
  %29 = tuple_extract %26 : $(Array<Int32>, Builtin.RawPointer), 1 // user: %31
  release_value %26 : $(Array<Int32>, Builtin.RawPointer) // id: %30
  %31 = pointer_to_address %29 : $Builtin.RawPointer to [strict] $*Int32 // users: %34, %36
  %32 = integer_literal $Builtin.Int32, 4         // user: %33
  %33 = struct $Int32 (%32 : $Builtin.Int32)      // user: %34
  store %33 to %31 : $*Int32                      // id: %34
  %35 = integer_literal $Builtin.Word, 1          // user: %36
  %36 = index_addr %31 : $*Int32, %35 : $Builtin.Word // user: %44
  %37 = metatype $@thin Int32.Type
  %38 = metatype $@thick Int32.Type               // user: %44
  %39 = integer_literal $Builtin.Int64, 1         // user: %40
  %40 = struct $Int (%39 : $Builtin.Int64)        // user: %42
  %41 = alloc_stack $Int                          // users: %42, %45, %44
  store %40 to %41 : $*Int                        // id: %42
  // function_ref SignedInteger<>.init<A>(_:)
  %43 = function_ref @$SSZss17FixedWidthIntegerRzrlEyxqd__cSzRd__lufC : $@convention(method) <τ_0_0 where τ_0_0 : FixedWidthInteger, τ_0_0 : SignedInteger><τ_1_0 where τ_1_0 : BinaryInteger> (@in τ_1_0, @thick τ_0_0.Type) -> @out τ_0_0 // user: %44
  %44 = apply %43<Int32, Int>(%36, %41, %38) : $@convention(method) <τ_0_0 where τ_0_0 : FixedWidthInteger, τ_0_0 : SignedInteger><τ_1_0 where τ_1_0 : BinaryInteger> (@in τ_1_0, @thick τ_0_0.Type) -> @out τ_0_0
  dealloc_stack %41 : $*Int                       // id: %45
```

lattner commented:
Oh interesting, thank you for tackling this.  I'm sorry to say this, but I have yet-another preferred way to fix this.  :-)  First, here's a slightly reduced testcase:

```
@inline(never)
func printTensor(_ x : Tensor<Float>) {
  print(x)
}
public func SR8399() {
  let x = Tensor<Float>(ones: [2, 2])
  let y = x.reshaped(toShape: Tensor<Int32>([4, Int32(1)]))
  printTensor(y)
}
```

The basic structural issue is that we've got more complex ways to initialize the array elements.  In this case, it is due to in-place initialization, but we're also not handling copy_addr and other things that can initialize values in decodeAllocUninitializedArray.

Instead of canonicalizing the array, I'd suggest just having decodeAllocUninitializedArray return information about the structure of the IR back to the client to handle.  decodeAllocUninitializedArray is used by both deabstraction of input lists (which only cares about the simple case) and by constexpr, which has more general logic that can already handle copy_addr and in-place init.

The specific change I'm imagining is that we change decodeAllocUninitializedArray to return a list of instruction*'s that initialize the elements (or just the address itself), replacing the 'SmallVectorImpl<SILValue> &elements' argument. 

The deabstraction client would just check for stores and fail on other cases.  The constexpr case would call computeSingleStoreAddressValue on it.

Does this make sense.  If you'd like, I can take care of this, but if you'd prefer to tackle it I'm totally +1 on that.  Thank you again for working on this @dingobye !

dingobye commented:
Thank you Mingsheng and Chris for the suggestions!

As for @mhong , it would be nice to be able to statically evaluate `SignedInteger<>.init<A>(_:)`. For this issue, the root cause seems not to be `SignedInteger<>.init<A>(_:)` itself but the address where it takes effects. Consider the following code:
```
  let x = Tensor<Float>(ones: [2, 2])
  let idx = Int32(1)
  let y = x.reshaped(toShape: Tensor<Int32>([4, idx]))
```
The SIL is something like:
```
%idx = alloc $Int32
call SignedInteger<>.init(%idx, ...)    // It writes to the stack variable rather than the 2nd elem's address in array
%top_level_value = load %idx
store %top_level_value to %addr_of_2nd_elem
```
Though it involves `SignedInteger<>.init<A>(_:)` as well, , `ConstExprEvaluator ` can handle this case because the element value corresponds is directly set by `store` and thus can be linked to a top level value.

As for @lattner , it definitely makes sense. It helps keep the IR unchanged in decoder. I would like to follow your advice and give it a try myself :-)

lattner commented:
Awesome, lemme know if you get stuck and I'm happy to help!  Thanks again!

dingobye committed:
[tensorflow] Handle in-place array element initialization
when decoding _allocateUninitialized.

This commit reverts the previous one which adds temporary
top-level stack variables. Instead, it adds supports to handle
in-place element initilization without changing IR when decoding
arrays.

dingobye commented:
I updated the code as per Chris' suggestions. The results of `decodeAllocUninitializedArray` is a list of `Operand*`, which can serve for both element address and initialiser purposes. I am not so happy with the parameter name `elementsAtInit`, though.

mhong commented on lib/SILOptimizer/Mandatory/TFConstExpr.cpp:
```diff
@@ -1528,18 +1545,20 @@ static bool analyzeArrayInitUses(SILValue v,
+/// function in the standard library.  This attempts to figure out how the
+/// resulting elements will be initialized.  This fills in the result with
+/// a lists of operands used to pass element addresses for initialization,
+/// and returns true on success.
```
return true on success --> should it be return false?

mhong commented on lib/SILOptimizer/Mandatory/TFConstExpr.cpp:
```diff
@@ -1601,20 +1621,34 @@ bool ConstExprEvaluator::decodeAllocUninitializedArray(
+        if (store->getDest() != use->get())
+          return true;
+      } else if (auto *applyInst = dyn_cast_or_null<ApplyInst>(user)) {
+        // Check to see if this is an out-parameter (i.e., like a store).
```
can we add a SIL snippet here as a code comment to show what cases we are dealing with here?

dingobye replied:
Added the `SignedInteger<>.init<A>(_:)` case

mhong reviewed (approved):
Looks good! Please also get LG from @lattner.

rxwei commented on lib/SILOptimizer/Mandatory/TFConstExpr.cpp:
```diff
@@ -1550,7 +1569,7 @@ bool ConstExprEvaluator::decodeAllocUninitializedArray(
 
   // The call has a tuple return type, see if we can walk all uses of them to
-  // analyze them and find the stores to the elements.
+  // analyze them and find the stores/apply_insts to the elements.
```
Nit: Replace "apply_insts" with just "apply's" here in comments? The use of snake case seems to refer to an instruction called `apply_inst` but it's actually `apply` :)

Same for other occurrences of "apply_inst".

dingobye committed:
Address comments from @mhong and @rxwei from PR-18389.

dingobye reviewed (commented):
Thank you Mingsheng and Richard! I rebased and updated code.

mhong commented:
Yes, pls feel free to replace print with _hostOp. Thanks for the clean-up work!

dingobye commented on test/TensorFlow/no_copy.swift:
```diff
@@ -184,14 +184,14 @@ struct Classifier {
   let loss = classifier.train(images: images, labels: labels,
                               learningRate: 0.3, epochCount: 100)
-  print(loss)
+  return loss
```
@mhong Is this change Okay? Because `loss` is `Float` and cannot be passed to `_hostOp`.

mhong replied:
ah right. we only have a version for tensors: https://github.com/apple/swift/blob/2b0fbd0be430cbcd89652e431c521a967f8daea3/stdlib/public/TensorFlow/Utilities.swift#L283

do we want to add a similar printScalar method, and mark it @inline(never)? One option is to put it in test utils in https://github.com/apple/swift/blob/2b0fbd0be430cbcd89652e431c521a967f8daea3/stdlib/private/TensorFlowUnittest/TensorFlowUnittest.swift. 

If there are more unexpected issues, please feel free to defer/skip such work. Thanks for the help!

dingobye replied:
Well, that makes sense. I'd like to do it in a separate PR later, since it is not directly relevant to this topic and may need to touch a number of test files.

rxwei replied:
Sure thing. Thanks!

dingobye replied:
I changed it back to `print` for this PR. Will clean them up later in a separate patch.

lattner commented:
Thanks, I'll take a look first thing tomorrow morning!

dingobye commented:
Thank you Chris! Please take your time.

mhong commented:
@swift-ci please test tensorflow

lattner commented:
Sorry for the delay, taking a look now.

lattner commented on lib/SILOptimizer/Mandatory/TFConstExpr.cpp:
```diff
@@ -1601,20 +1621,47 @@ bool ConstExprEvaluator::decodeAllocUninitializedArray(
+      elementsAtInit[index] = use;
+
+      if (arrayInsts)
+        arrayInsts->insert(user);
```
in the apply case, please set 'hadUnknownUsers = true' just to be conservatively correct.  We can further optimize this in the future.

dingobye replied:
Done

lattner reviewed (approved):
Minor comments above, but otherwise looks really great.  Thank you so much for fixing this!!

dingobye committed:
Address comments from @lattner from #18389.

dingobye commented on lib/SILOptimizer/Mandatory/TFConstExpr.cpp:
```diff
@@ -814,19 +814,28 @@ ConstExprFunctionState::computeCallResult(ApplyInst *apply) {
-      for (auto elt : elements) {
-        auto eltCst = getConstantValue(elt);
+      for (auto *use : elementsAtInit) {
+        SymbolicValue eltCst = evaluator.getUnknown((SILInstruction *)apply,
```
Makes sense :-)

dingobye commented on lib/SILOptimizer/Mandatory/TFConstExpr.cpp:
```diff
@@ -814,19 +814,28 @@ ConstExprFunctionState::computeCallResult(ApplyInst *apply) {
+        } else {
+          auto addr = computeSingleStoreAddressValue(use->get());
+          if (addr.getKind() == SymbolicValue::Address)
+            eltCst = addr.getAddressValueMemoryObject()->getValue();
```
Yes, it works! `computeLoadResult` is actually much better here. Because `computeSingleStoreAddressValue` is supposed to accept `alloc`s only; if we want to pass a non-`alloc` address to it, we'd need to modify its implementation to break such restriction.

dingobye commented:
Thank you for the comments, Chris! I updated the code as you suggested.

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

lattner commented:
Awesome, thank you again!
